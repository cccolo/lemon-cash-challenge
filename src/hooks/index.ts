export {useAuth} from './useAuth';
export {useCryptocurrency} from './useCryptocurrency';
export {useCryptocurrencySearch} from './useCryptocurrencySearch';
export {useCryptocurrencyFavorites} from './useCryptocurrencyFavorites';
export {useCryptocurrencyFavoriteSearch} from './useCryptocurrencyFavoriteSearch';
