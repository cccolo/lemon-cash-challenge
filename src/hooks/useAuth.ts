import * as React from 'react';
import {
  statusCodes,
  isErrorWithCode,
  GoogleSignin,
} from '@react-native-google-signin/google-signin';
import {GOOGLE_SIGN_IN_WEB_CLIENT_ID} from '@env';
import {AuthContext} from '../context/AuthContext';

GoogleSignin.configure({
  webClientId: GOOGLE_SIGN_IN_WEB_CLIENT_ID,
  offlineAccess: true,
});

export const useAuth = () => {
  const context = React.useContext(AuthContext);

  const googleSignIn = async () => {
    try {
      await GoogleSignin.hasPlayServices({showPlayServicesUpdateDialog: true});
      const {idToken} = await GoogleSignin.signIn();
      return idToken || '';
    } catch (error) {
      if (isErrorWithCode(error)) {
        switch (error.code) {
          case statusCodes.SIGN_IN_CANCELLED:
            // user cancelled the login flow
            break;
          case statusCodes.IN_PROGRESS:
            // operation (eg. sign in) already in progress
            break;
          case statusCodes.PLAY_SERVICES_NOT_AVAILABLE:
            // play services not available or outdated
            break;
          default:
          // some other error happened
        }
      } else {
        // an error that's not related to google sign in occurred
      }
    }
  };

  const signIn = async () => {
    const token = await googleSignIn();
    if (!token) {
      return;
    }
    context?.signIn(token);
  };

  const signOff = () => {
    context?.signOff();
  };

  return {
    signIn,
    signOff,
  };
};
