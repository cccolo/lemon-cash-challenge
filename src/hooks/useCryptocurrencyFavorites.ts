import React from 'react';
import {setData, getData} from '../commons/asyncStorage';

interface Favorites {
  [key: number]: number;
}

const KEY = 'cryptocurrencyFav';

export const useCryptocurrencyFavorites = () => {
  const [favorites, setFavorites] = React.useState<Favorites>({});

  const setFavoriteItems = React.useCallback(
    (id: number) => {
      if (favorites[id]) {
        const newFavorites = {...favorites};
        delete newFavorites[id];
        setData(KEY, Object.values(newFavorites).join(','));
        setFavorites(newFavorites);
        return;
      }

      const newFavorites = {...favorites, [id]: id};
      setData(KEY, Object.values(newFavorites).join(','));
      setFavorites(newFavorites);
    },
    [favorites],
  );

  React.useEffect(() => {
    const init = async () => {
      const cryptocurrencyFavFromStorage = await getData(KEY);
      const cryptocurrencyFavToArray = cryptocurrencyFavFromStorage?.split(',');
      if (cryptocurrencyFavToArray?.length) {
        const cryptocurrencyFav: Favorites = cryptocurrencyFavToArray.reduce(
          (prev: Favorites, current: any) => {
            prev[current] = current;
            return prev;
          },
          {},
        );
        setFavorites(cryptocurrencyFav);
      }
    };
    init();
  }, []);

  const favoritesToString = async () => {
    return (await getData(KEY)) || '';
  };

  const clearFavoriteItems = async () => {
    return await setData(KEY, '');
  };

  return {
    setFavoriteItems,
    favorites,
    favoritesToString,
    clearFavoriteItems,
  };
};
