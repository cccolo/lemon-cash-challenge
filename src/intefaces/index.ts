export type {Cryptocurrency, CryptocurrencyData} from './cryptocurrency';
