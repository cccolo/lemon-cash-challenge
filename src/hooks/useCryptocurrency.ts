import {useInfiniteQuery} from '@tanstack/react-query';
import {coinMarketCapAPi} from '../api/coinMarketCapAPi';
import {Cryptocurrency} from '../intefaces';

interface QueryProps {
  pageParam: number;
}

const LIMIT_SIZE = 20;

const getCryptocurrencyList = async ({
  pageParam,
}: QueryProps): Promise<Cryptocurrency> => {
  const {data} = await coinMarketCapAPi.get<Cryptocurrency>(
    `/v1/cryptocurrency/listings/latest?start=${
      pageParam * LIMIT_SIZE
    }&limit=${LIMIT_SIZE}`,
  );

  return data;
};

export const useCryptocurrency = () => {
  const cryptocurrencyInfiniteQuery = useInfiniteQuery({
    queryKey: ['cryptocurrency', 'listings'],
    queryFn: getCryptocurrencyList,
    initialPageParam: 1,
    getNextPageParam: (lastPage, pages) => {
      if (lastPage?.data?.length === 0) {
        return;
      }
      return pages.length + 1;
    },
  });

  return {
    cryptocurrencyInfiniteQuery,
  };
};
