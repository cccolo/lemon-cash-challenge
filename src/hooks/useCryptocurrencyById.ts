import {useQuery} from '@tanstack/react-query';
import {coinMarketCapAPi} from '../api/coinMarketCapAPi';
import {Cryptocurrency} from '../intefaces';

const getCryptocurrencyById = async (
  value: string | number,
): Promise<Cryptocurrency> => {
  const {data} = await coinMarketCapAPi.get<Cryptocurrency>(
    `v2/cryptocurrency/quotes/latest?id=${value}`,
  );

  return data;
};

export const useCryptocurrencyById = (
  value: string | number,
  enabled = false,
  refetchIntervalEnabled = false,
) => {
  const cryptocurrencyByIdQuery = useQuery({
    queryKey: ['cryptocurrency', 'listings', 'id', value],
    queryFn: async () => await getCryptocurrencyById(value),
    enabled: enabled && !!value,
    refetchInterval: refetchIntervalEnabled && 30000,
  });

  return {
    cryptocurrencyByIdQuery,
  };
};
