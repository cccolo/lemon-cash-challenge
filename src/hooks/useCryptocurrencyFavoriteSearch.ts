import {useQuery} from '@tanstack/react-query';
import {coinMarketCapAPi} from '../api/coinMarketCapAPi';
import {Cryptocurrency} from '../intefaces';

const getCryptocurrencyById = async (
  value: string,
): Promise<Cryptocurrency> => {
  const {data} = await coinMarketCapAPi.get<Cryptocurrency>(
    `v2/cryptocurrency/quotes/latest?id=${value}`,
  );

  return data;
};

export const useCryptocurrencyFavoriteSearch = (
  value: string,
  enabled = false,
) => {
  const cryptocurrencyFavoriteQuery = useQuery({
    queryKey: ['cryptocurrency', 'listings', 'id', value],
    queryFn: async () => await getCryptocurrencyById(value),
    enabled: enabled && !!value,
  });

  return {
    cryptocurrencyFavoriteQuery,
  };
};
