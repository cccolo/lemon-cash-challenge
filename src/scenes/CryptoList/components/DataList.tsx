import React from 'react';
import {
  ActivityIndicator,
  FlatList,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';
import {CryptocurrencyData} from '../../../intefaces';
import {StartIcon, StartSelectIcon} from '../../../assets/images';
import {useCryptocurrencyFavorites} from '../../../hooks';

type Props = {
  data: CryptocurrencyData[];
  fetchNextPage?: () => void;
  hasPagination?: boolean;
  isLoading: boolean;
};

export const DataList: React.FC<Props> = ({
  isLoading,
  fetchNextPage,
  hasPagination,
  data,
}: Props) => {
  const {setFavoritesItems, favorites} = useCryptocurrencyFavorites();

  const cryptocurrencyItemExtractorKey = React.useCallback(
    (item: CryptocurrencyData) => {
      return item.id.toString();
    },
    [],
  );

  const fetchNext = React.useCallback(() => {
    if (!hasPagination) {
      return;
    }
    fetchNextPage?.();
  }, [hasPagination, fetchNextPage]);

  const ListEndLoader = React.useCallback(() => {
    if (isLoading) {
      return (
        <View style={styles.loader}>
          <ActivityIndicator size={'large'} />
        </View>
      );
    }
  }, [isLoading]);

  const renderItem = React.useCallback(
    (item: CryptocurrencyData) => {
      return (
        <View style={styles.container}>
          <View>
            <Text style={[styles.description, styles.symbol]}>
              {item.symbol}
            </Text>
            <View>
              <Text style={[styles.description, styles.name]}>{item.name}</Text>
            </View>
          </View>
          <View style={styles.price}>
            <Text style={styles.description}>{item.quote.USD.price}</Text>
          </View>
          <View style={styles.favorites}>
            <TouchableOpacity onPress={() => setFavoritesItems(item.id)}>
              {favorites[item.id] ? <StartSelectIcon /> : <StartIcon />}
            </TouchableOpacity>
          </View>
        </View>
      );
    },
    [favorites, setFavoritesItems],
  );

  return (
    <FlatList
      data={data}
      onEndReached={fetchNext}
      onEndReachedThreshold={0.8}
      ListFooterComponent={ListEndLoader}
      renderItem={({item}) => renderItem(item)}
      keyExtractor={cryptocurrencyItemExtractorKey}
    />
  );
};

// TODO:
const styles = StyleSheet.create({
  container: {
    borderColor: 'rgba(255, 255, 255, 0.4)',
    borderRadius: 5,
    borderWidth: 1,
    flex: 1,
    flexDirection: 'row',
    marginBottom: 5,
    marginTop: 5,
    padding: 8,
  },
  symbol: {
    fontFamily: 'NeueMachina-Ultrabold',
    textAlign: 'left',
    fontSize: 14,
  },
  name: {
    color: 'white',
    fontSize: 12,
    textAlign: 'left',
  },
  price: {
    flex: 1,
    justifyContent: 'center',
    paddingRight: 10,
  },
  loader: {
    margin: 10,
  },
  description: {
    color: 'rgb(0, 240, 104)',
    fontFamily: 'NeueMachina-Regular',
    textAlign: 'right',
    lineHeight: 16,
    fontSize: 12,
  },
  favorites: {
    justifyContent: 'center',
  },
});
