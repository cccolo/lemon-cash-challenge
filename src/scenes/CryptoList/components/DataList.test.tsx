import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import {DataList} from './DataList';
import {CryptocurrencyData} from '../../../intefaces';
import {useCryptocurrencyFavorites} from '../../../hooks';

// Mocking the custom hook
jest.mock('../../../hooks', () => ({
  useCryptocurrencyFavorites: jest.fn(),
}));

// Mocking the StartIcon and StartSelectIcon components
jest.mock('../../../assets/images', () => ({
  StartIcon: () => <></>,
  StartSelectIcon: () => <></>,
}));

const mockUseCryptocurrencyFavorites =
  useCryptocurrencyFavorites as jest.MockedFunction<
    typeof useCryptocurrencyFavorites
  >;

describe('DataList Component', () => {
  const data: CryptocurrencyData[] = [
    {
      id: 1,
      name: 'Bitcoin',
      symbol: 'BTC',
      slug: 'bitcoin',
      num_market_pairs: 10,
      date_added: new Date('2024-06-12T14:58:53.329Z'),
      tags: ['tag1', 'tag2'],
      max_supply: 21000000,
      circulating_supply: 19000000,
      total_supply: 20000000,
      infinite_supply: false,
      platform: null,
      cmc_rank: 1,
      self_reported_circulating_supply: null,
      self_reported_market_cap: null,
      tvl_ratio: null,
      last_updated: new Date('2024-06-12T14:58:53.329Z'),
      quote: {
        USD: {
          price: 30000,
          volume_24h: 1000000,
          volume_change_24h: 5,
          percent_change_1h: 0.1,
          percent_change_24h: 1,
          percent_change_7d: 7,
          percent_change_30d: 30,
          percent_change_60d: 60,
          percent_change_90d: 90,
          market_cap: 600000000,
          market_cap_dominance: 40,
          fully_diluted_market_cap: 630000000,
          tvl: null,
          last_updated: new Date('2024-06-12T14:58:53.329Z'),
        },
      },
    },
    {
      id: 2,
      name: 'Ethereum',
      symbol: 'ETH',
      slug: 'ethereum',
      num_market_pairs: 20,
      date_added: new Date('2024-06-12T14:58:53.329Z'),
      tags: ['tag3', 'tag4'],
      max_supply: null,
      circulating_supply: 110000000,
      total_supply: 120000000,
      infinite_supply: true,
      platform: null,
      cmc_rank: 2,
      self_reported_circulating_supply: null,
      self_reported_market_cap: null,
      tvl_ratio: null,
      last_updated: new Date('2024-06-12T14:58:53.329Z'),
      quote: {
        USD: {
          price: 2000,
          volume_24h: 500000,
          volume_change_24h: 3,
          percent_change_1h: 0.05,
          percent_change_24h: 0.5,
          percent_change_7d: 5,
          percent_change_30d: 20,
          percent_change_60d: 50,
          percent_change_90d: 80,
          market_cap: 220000000,
          market_cap_dominance: 20,
          fully_diluted_market_cap: 230000000,
          tvl: null,
          last_updated: new Date('2024-06-12T14:58:53.329Z'),
        },
      },
    },
  ];

  const setFavoriteItemsMock = jest.fn();
  const onSelectItemMock = jest.fn();
  const fetchNextPageMock = jest.fn();

  beforeEach(() => {
    mockUseCryptocurrencyFavorites.mockReturnValue({
      favorites: {1: 1},
      setFavoriteItems: setFavoriteItemsMock,
      favoritesToString: jest.fn(),
      clearFavoriteItems: jest.fn(),
    });
  });

  it('renders correctly and matches snapshot', () => {
    const {toJSON} = render(
      <DataList
        data={data}
        isLoading={false}
        fetchNextPage={fetchNextPageMock}
        hasPagination={true}
        onSelectItem={onSelectItemMock}
      />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders loading indicator when isLoading is true', () => {
    const {getByTestId} = render(
      <DataList
        data={[]}
        isLoading={true}
        fetchNextPage={fetchNextPageMock}
        hasPagination={true}
        onSelectItem={onSelectItemMock}
      />,
    );

    expect(getByTestId('activity-indicator')).toBeTruthy();
  });

  it('calls onSelectItem when an item is pressed', () => {
    const {getByTestId} = render(
      <DataList
        data={data}
        isLoading={false}
        fetchNextPage={fetchNextPageMock}
        hasPagination={true}
        onSelectItem={onSelectItemMock}
      />,
    );

    fireEvent.press(getByTestId('item-1'));
    expect(onSelectItemMock).toHaveBeenCalledWith(1);
  });

  it('calls setFavoriteItems when favorite icon is pressed', () => {
    const {getByTestId} = render(
      <DataList
        data={data}
        isLoading={false}
        fetchNextPage={fetchNextPageMock}
        hasPagination={true}
        onSelectItem={onSelectItemMock}
      />,
    );

    fireEvent.press(getByTestId('favorite-button-1'));
    expect(setFavoriteItemsMock).toHaveBeenCalledWith(1);
  });
});
