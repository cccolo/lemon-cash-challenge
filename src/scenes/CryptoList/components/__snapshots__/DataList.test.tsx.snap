// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataList Component renders correctly and matches snapshot 1`] = `
<RCTScrollView
  ListEmptyComponent={[Function]}
  ListFooterComponent={[Function]}
  data={
    [
      {
        "circulating_supply": 19000000,
        "cmc_rank": 1,
        "date_added": 2024-06-12T14:58:53.329Z,
        "id": 1,
        "infinite_supply": false,
        "last_updated": 2024-06-12T14:58:53.329Z,
        "max_supply": 21000000,
        "name": "Bitcoin",
        "num_market_pairs": 10,
        "platform": null,
        "quote": {
          "USD": {
            "fully_diluted_market_cap": 630000000,
            "last_updated": 2024-06-12T14:58:53.329Z,
            "market_cap": 600000000,
            "market_cap_dominance": 40,
            "percent_change_1h": 0.1,
            "percent_change_24h": 1,
            "percent_change_30d": 30,
            "percent_change_60d": 60,
            "percent_change_7d": 7,
            "percent_change_90d": 90,
            "price": 30000,
            "tvl": null,
            "volume_24h": 1000000,
            "volume_change_24h": 5,
          },
        },
        "self_reported_circulating_supply": null,
        "self_reported_market_cap": null,
        "slug": "bitcoin",
        "symbol": "BTC",
        "tags": [
          "tag1",
          "tag2",
        ],
        "total_supply": 20000000,
        "tvl_ratio": null,
      },
      {
        "circulating_supply": 110000000,
        "cmc_rank": 2,
        "date_added": 2024-06-12T14:58:53.329Z,
        "id": 2,
        "infinite_supply": true,
        "last_updated": 2024-06-12T14:58:53.329Z,
        "max_supply": null,
        "name": "Ethereum",
        "num_market_pairs": 20,
        "platform": null,
        "quote": {
          "USD": {
            "fully_diluted_market_cap": 230000000,
            "last_updated": 2024-06-12T14:58:53.329Z,
            "market_cap": 220000000,
            "market_cap_dominance": 20,
            "percent_change_1h": 0.05,
            "percent_change_24h": 0.5,
            "percent_change_30d": 20,
            "percent_change_60d": 50,
            "percent_change_7d": 5,
            "percent_change_90d": 80,
            "price": 2000,
            "tvl": null,
            "volume_24h": 500000,
            "volume_change_24h": 3,
          },
        },
        "self_reported_circulating_supply": null,
        "self_reported_market_cap": null,
        "slug": "ethereum",
        "symbol": "ETH",
        "tags": [
          "tag3",
          "tag4",
        ],
        "total_supply": 120000000,
        "tvl_ratio": null,
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onEndReachedThreshold={0.8}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={[]}
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="item-1"
      >
        <View
          style={
            {
              "borderColor": "rgba(255, 255, 255, 0.4)",
              "borderRadius": 5,
              "borderWidth": 1,
              "flex": 1,
              "flexDirection": "row",
              "marginBottom": 5,
              "marginTop": 5,
              "padding": 8,
            }
          }
        >
          <View>
            <Text
              style={
                [
                  {
                    "color": "rgb(0, 240, 104)",
                    "fontFamily": "NeueMachina-Regular",
                    "fontSize": 12,
                    "lineHeight": 16,
                    "textAlign": "right",
                  },
                  {
                    "fontFamily": "NeueMachina-Ultrabold",
                    "fontSize": 14,
                    "textAlign": "left",
                  },
                ]
              }
            >
              BTC
            </Text>
            <View>
              <Text
                style={
                  [
                    {
                      "color": "rgb(0, 240, 104)",
                      "fontFamily": "NeueMachina-Regular",
                      "fontSize": 12,
                      "lineHeight": 16,
                      "textAlign": "right",
                    },
                    {
                      "color": "rgb(255, 255, 255)",
                      "fontSize": 12,
                      "textAlign": "left",
                    },
                  ]
                }
              >
                Bitcoin
              </Text>
            </View>
          </View>
          <View
            style={
              {
                "flex": 1,
                "justifyContent": "center",
                "paddingRight": 10,
              }
            }
          >
            <Text
              style={
                {
                  "color": "rgb(0, 240, 104)",
                  "fontFamily": "NeueMachina-Regular",
                  "fontSize": 12,
                  "lineHeight": 16,
                  "textAlign": "right",
                }
              }
            >
              30000
            </Text>
          </View>
          <View
            style={
              {
                "justifyContent": "center",
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              hitSlop={
                {
                  "bottom": 10,
                  "left": 10,
                  "right": 10,
                  "top": 10,
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
              testID="favorite-button-1"
            />
          </View>
        </View>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="item-2"
      >
        <View
          style={
            {
              "borderColor": "rgba(255, 255, 255, 0.4)",
              "borderRadius": 5,
              "borderWidth": 1,
              "flex": 1,
              "flexDirection": "row",
              "marginBottom": 5,
              "marginTop": 5,
              "padding": 8,
            }
          }
        >
          <View>
            <Text
              style={
                [
                  {
                    "color": "rgb(0, 240, 104)",
                    "fontFamily": "NeueMachina-Regular",
                    "fontSize": 12,
                    "lineHeight": 16,
                    "textAlign": "right",
                  },
                  {
                    "fontFamily": "NeueMachina-Ultrabold",
                    "fontSize": 14,
                    "textAlign": "left",
                  },
                ]
              }
            >
              ETH
            </Text>
            <View>
              <Text
                style={
                  [
                    {
                      "color": "rgb(0, 240, 104)",
                      "fontFamily": "NeueMachina-Regular",
                      "fontSize": 12,
                      "lineHeight": 16,
                      "textAlign": "right",
                    },
                    {
                      "color": "rgb(255, 255, 255)",
                      "fontSize": 12,
                      "textAlign": "left",
                    },
                  ]
                }
              >
                Ethereum
              </Text>
            </View>
          </View>
          <View
            style={
              {
                "flex": 1,
                "justifyContent": "center",
                "paddingRight": 10,
              }
            }
          >
            <Text
              style={
                {
                  "color": "rgb(0, 240, 104)",
                  "fontFamily": "NeueMachina-Regular",
                  "fontSize": 12,
                  "lineHeight": 16,
                  "textAlign": "right",
                }
              }
            >
              2000
            </Text>
          </View>
          <View
            style={
              {
                "justifyContent": "center",
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              hitSlop={
                {
                  "bottom": 10,
                  "left": 10,
                  "right": 10,
                  "top": 10,
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
              testID="favorite-button-2"
            />
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
    />
  </View>
</RCTScrollView>
`;
